### PR TITLE
More ProposerPriority tests 

### PR DESCRIPTION
**Pull request by Liamsi**
* [ ] Updated all relevant documentation in docs
* [ ] Updated all code comments where relevant
* [ ] Wrote tests
* [ ] Updated CHANGELOG_PENDING.md

### CODE REVIEW

File: state/state_test.go [hunk 1]
```diff
@@ -263,6 +263,294 @@ func TestOneValidatorChangesSaveLoad(t *testing.T) {
...
+	validatorUpdates, err = types.PB2TM.ValidatorUpdates([]abci.ValidatorUpdate{updateAddVal})
+	assert.NoError(t, err)
+	updatedState2, err := updateState(updatedState, blockID, &block.Header, abciResponses, validatorUpdates)
+	assert.NoError(t, err)
+
+	require.Equal(t, len(updatedState2.NextValidators.Validators), 2)
+	_, addedVal2 := updatedState2.NextValidators.GetByAddress(val2PubKey.Address())
+	// adding a validator should not lead to a ProposerPriority equal to zero (unless the combination of averaging and
+	// incrementing would cause so; which is not the case here)
+	assert.NotZero(t, addedVal2.ProposerPriority, "Added validator's ProposerPriority should not be zero")
```
Comment (cwgoes) [inline state/state_test.go]: Shouldn't we know what it would be exactly - can we just assert that?

File: state/state_test.go [hunk 2]
```diff
@@ -263,6 +263,294 @@ func TestOneValidatorChangesSaveLoad(t *testing.T) {
...
+	// Updating a validator does not reset the ProposerPriority to zero:
+	updateVal := abci.ValidatorUpdate{PubKey: types.TM2PB.PubKey(val2PubKey), Power: 1}
+	validatorUpdates, err = types.PB2TM.ValidatorUpdates([]abci.ValidatorUpdate{updateVal})
+	assert.NoError(t, err)
+	updatedState3, err := updateState(updatedState2, blockID, &block.Header, abciResponses, validatorUpdates)
+	assert.NoError(t, err)
+
+	require.Equal(t, len(updatedState3.NextValidators.Validators), 2)
+	_, updatedVal2 := updatedState3.NextValidators.GetByAddress(val2PubKey.Address())
+	assert.NotZero(t, updatedVal2.ProposerPriority)
```
Comment (cwgoes) [inline state/state_test.go]: Ditto, why not assert exact?

File: state/state_test.go [hunk 3]
```diff
@@ -263,6 +263,294 @@ func TestOneValidatorChangesSaveLoad(t *testing.T) {
...
+
+	// val1 will still be proposer as val2 just got added:
+	assert.Equal(t, updatedState2.Validators.Proposer.Address, updatedState2.NextValidators.Proposer.Address)
+	assert.Equal(t, updatedState2.Validators.Proposer.Address, val1PubKey.Address())
+	assert.Equal(t, updatedState2.NextValidators.Proposer.Address, val1PubKey.Address())
+
+	_, updatedVal1 := updatedState2.NextValidators.GetByAddress(val1PubKey.Address())
+	_, updatedVal2 := updatedState2.NextValidators.GetByAddress(val2PubKey.Address())
+	// val1's ProposerPriority story: -10 (see above) + 10 (voting pow) - (-1) (avg) - 20 (total) == -19
+	assert.EqualValues(t, -19, updatedVal1.ProposerPriority)
```
Comment (cwgoes) [inline state/state_test.go]: Can we calculate `-19` in terms of the initial priorities (inline)?

File: state/state_test.go [hunk 4]
```diff
@@ -263,6 +263,294 @@ func TestOneValidatorChangesSaveLoad(t *testing.T) {
...
+	// val2 prio when added: -(totalVotingPower + (totalVotingPower >> 3)) == -11
+	// -> -11 + 10  (voting power) - (-1) (avg) == 0
+	assert.EqualValues(t, 0, updatedVal2.ProposerPriority, "unexpected proposer priority for validator: %v", updatedVal2)
```
Comment (cwgoes) [inline state/state_test.go]: Can we calculate `0` in terms of the initial priorities (inline)?

File: state/state_test.go [hunk 5]
```diff
@@ -263,6 +263,294 @@ func TestOneValidatorChangesSaveLoad(t *testing.T) {
...
+
+	block := makeBlock(state, state.LastBlockHeight+1)
+	blockID := types.BlockID{block.Hash(), block.MakePartSet(testPartSize).Header()}
+	// no updates:
+	abciResponses := &ABCIResponses{
+		EndBlock: &abci.ResponseEndBlock{ValidatorUpdates: nil},
+	}
+	validatorUpdates, err := types.PB2TM.ValidatorUpdates(abciResponses.EndBlock.ValidatorUpdates)
+	require.NoError(t, err)
+
```
Comment (cwgoes) [inline state/state_test.go]: Can we also assert which validator is the proposer (in all of these tests)?

File: state/state_test.go [hunk 6]
```diff
@@ -263,6 +263,294 @@ func TestOneValidatorChangesSaveLoad(t *testing.T) {
...
+	assert.NotEqual(t, updatedState3.Validators.Proposer.Address, updatedState3.NextValidators.Proposer.Address)
+
+	assert.Equal(t, updatedState3.Validators, updatedState2.NextValidators)
+	_, updatedVal1 = updatedState3.NextValidators.GetByAddress(val1PubKey.Address())
+	_, updatedVal2 = updatedState3.NextValidators.GetByAddress(val2PubKey.Address())
+
+	// check if expected proposer prio is matched:
+
+	// -19 + 10 - 0 (avg) == -9
+	assert.EqualValues(t, -9, updatedVal1.ProposerPriority, "unexpected proposer priority for validator: %v", updatedVal2)
```
Comment (cwgoes) [inline state/state_test.go]: Ditto on above (clculate inline)

File: state/state_test.go [hunk 7]
```diff
@@ -263,6 +263,294 @@ func TestOneValidatorChangesSaveLoad(t *testing.T) {
...
+		// no validator updates:
+		abciResponses := &ABCIResponses{
+			EndBlock: &abci.ResponseEndBlock{ValidatorUpdates: nil},
+		}
+		validatorUpdates, err = types.PB2TM.ValidatorUpdates(abciResponses.EndBlock.ValidatorUpdates)
+		require.NoError(t, err)
+
+		updatedState, err := updateState(oldState, blockID, &block.Header, abciResponses, validatorUpdates)
+		assert.NoError(t, err)
+		// alternate:
```
Comment (cwgoes) [inline state/state_test.go]: cool! can we also check that the powers change cyclically (as they ought to)?

File: state/state_test.go [hunk 8]
```diff
@@ -263,6 +263,294 @@ func TestOneValidatorChangesSaveLoad(t *testing.T) {
...
+		block = makeBlock(curState, curState.LastBlockHeight+1)
+		blockID = types.BlockID{block.Hash(), block.MakePartSet(testPartSize).Header()}
+		curState, err = updateState(curState, blockID, &block.Header, abciResponses, validatorUpdates)
+		if !bytes.Equal(curState.Validators.Proposer.Address, curState.NextValidators.Proposer.Address) {
+			isProposerUnchanged = false
+		}
+		count++
+	}
+	// first proposer change happens after this many iters; we probably want to lower this number:
+	// TODO: change with https://github.com/tendermint/tendermint/issues/2960
```
Comment (cwgoes) [inline state/state_test.go]: cool test, ideally we could calculate what this value ought to be in terms of the power ratio, and then of course try to fix it

Comment (Liamsi) [reply state/state_test.go]: I think, I've addressed all your comments besides this one. I'll use this to see if we can change the behaviour instead. Please take another look if you like.

**Comment by cwgoes** [comment]
The testcase additions look good to me.

After reading and reasoning through the proposer election mechanism in detail with @Liamsi, I think it would greatly benefit in clarity (and in lowering the likelihood of mistakes in implementation) from rearchitecting for clarity. A lot of the code does not make too much sense - e.g. we create a heap each round to only calculate a maximum, which is much slower than just iterating through the validators once, pass around by reference or by copy in inconsistent ways, shift by the average more often than necessary (still, I think, even after the once-every-n - I think we only need to shift by the average once).

**Comment by Liamsi** [comment]
> we create a heap each round to only calculate a maximum, which is much slower than just iterating through the validators once,

I'm removing the heap in the follow up PR for #2960. 

Thanks a lot @cwgoes!!!

**Pull request merged**